Source/Logger/trace.cpp

```diff
@@ -0,0 +1,304 @@
+
+template<size_t SIZE>
+int stprintf_s(wchar_t(&buffer)[SIZE], _Printf_format_string_ wchar_t const* format, ...)
+{
+    va_list varArgs = nullptr;
+    va_start(varArgs, format);
+    auto result = vswprintf_s(buffer, format, varArgs);
+    va_end(varArgs);
+    return result;
+}
+
+int stprintf_s(wchar_t* buffer, size_t size, _Printf_format_string_ wchar_t const* format, ...)
+{
+    va_list varArgs = nullptr;
+    va_start(varArgs, format);
+    auto result = vswprintf_s(buffer, size, format, varArgs);
+    va_end(varArgs);
+    return result;
+}
+
+template<size_t SIZE>
+int vstprintf_s(wchar_t(&buffer)[SIZE], _Printf_format_string_ wchar_t const* format, va_list varArgs)
+{
+    return vswprintf_s(buffer, format, varArgs);
+}
+
+void OutputDebugStringT(wchar_t const* string)
+{
+    OutputDebugStringW(string);
+}
+
+#else
+
+template<size_t SIZE>
+int stprintf_s(char(&buffer)[SIZE], _Printf_format_string_ char const* format ...)
+{
+    va_list varArgs = nullptr;
+    va_start(varArgs, format);
+    auto result = vsprintf_s(buffer, format, varArgs);
+    va_end(varArgs);
+    return result;
+}
+
+int stprintf_s(char* buffer, size_t size, _Printf_format_string_ char const* format ...)
+{
+    va_list varArgs = nullptr;
+    va_start(varArgs, format);
+    auto result = vsprintf_s(buffer, size, format, varArgs);
+    va_end(varArgs);
+    return result;
+}
+
+template<size_t SIZE>
+int vstprintf_s(char(&buffer)[SIZE], _Printf_format_string_ char const* format, va_list varArgs)
+{
+    return vsprintf_s(buffer, format, varArgs);
+}
+
+void OutputDebugStringT(char const* string)
+{
+    OutputDebugStringA(string);
+}
+
+#endif
+
+//------------------------------------------------------------------------------
+// Trace implementation
+//------------------------------------------------------------------------------
+class TraceState
+{
+public:
+    void Init()
+    {
+        auto previousCount = m_tracingClients.fetch_add(1);
+        if (previousCount == 0)
+        {
+            m_initTime = std::chrono::high_resolution_clock::now();
+        }
+    }
+
+    void Cleanup()
+    {
+        --m_tracingClients;
+    }
+
+    bool IsSetup() const
+    {
+        return m_tracingClients > 0;
+    }
+
+    void SetClientCallback(HCTraceCallback* callback)
+    {
+        m_clientCallback = callback;
+    }
+
+    HCTraceCallback* GetClientCallback() const
+    {
+        return m_clientCallback;
+    }
+
+    uint64_t GetTimestamp() const
+    {
+        auto now = std::chrono::high_resolution_clock::now();
+        auto nowMS = std::chrono::duration_cast<std::chrono::milliseconds>(now - m_initTime.load());
+        return nowMS.count();
+    }
+
+private:
+    std::atomic<uint32_t> m_tracingClients = 0;
+    std::atomic<std::chrono::high_resolution_clock::time_point> m_initTime =
+        std::chrono::high_resolution_clock::now();
+    std::atomic<HCTraceCallback*> m_clientCallback = nullptr;
+};
+
+TraceState g_traceState;
```

Comment:
global state.  can we avoid and put it in the singleton that's created upon hcglobalinit()?

Reply:
I was trying to avoid having a dependency from the logging code to the rest of LibHttpClient, which is why it has its own global init and cleanup.

The main reason for this is so that it can be used by other consumers without bringing in the rest of LibHttpClient (if they want)